File: web/src/types/events.ts
/**
 * Typed WebSocket event discriminated union.
 *
 * Mirrors server/events.py dataclass shapes.  The 'type' field is the
 * discriminator — handler maps use Extract<> for type narrowing.
 *
 * Pattern: CorvidAgent shared type layer + react-socket typed events.
 * Source: https://github.com/CorvidLabs/corvid-agent/issues/957
 */

// ── EventEnvelope base ────────────────────────────────────────────────────

/** Fields injected by the backend on every WS event. All optional for backward compat. */
export interface EventEnvelope {
  session_id?: string;
  run_id?: string;
  turn_id?: string;
  event_id?: string;
  sequence?: number;
  block_id?: string;
  tool_call_id?: string;
  turn_index?: number;
}

// ── Status ──────────────────────────────────────────────────────────────

export interface WsStatusEvent extends EventEnvelope {
  type: "status";
  status: "running" | "completed" | "failed" | "finish" | "gave_up" | "cancelled" | "compacted";
  message?: string;
  error?: string;
  result?: { summary?: string; steps_taken?: number; total_tokens?: number };
  content?: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

// ── Thought / Reflection ────────────────────────────────────────────────

export interface WsThoughtEvent extends EventEnvelope {
  type: "thought";
  content: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

export interface WsThoughtDeltaEvent extends EventEnvelope {
  type: "thought_delta";
  text: string;
  timestamp?: string;
  step?: number;
  child_session_id?: string;
}

export interface WsReflectionEvent extends EventEnvelope {
  type: "reflection";
  content: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
}

// ── Tool call / Observation ─────────────────────────────────────────────

export interface WsToolCallEvent extends EventEnvelope {
  type: "tool_call";
  name: string;
  params?: Record<string, unknown>;
  id?: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

export interface WsObservationEvent extends EventEnvelope {
  type: "observation";
  tool_name?: string;
  output?: string;
  error?: string;
  status?: string;
  id?: string;
  paired?: boolean;
  diff?: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

// ── Subagent ────────────────────────────────────────────────────────────

export interface WsSubagentStartEvent extends EventEnvelope {
  type: "subagent_start";
  child_session_id: string;
  agent_name?: string;
  timestamp?: string;
  step?: number;
}

export interface WsSubagentStopEvent extends EventEnvelope {
  type: "subagent_stop";
  child_session_id: string;
  status?: string;
  timestamp?: string;
  step?: number;
}

// ── Multi-agent delegation ─────────────────────────────────────────────

export interface DelegationBudgetPayload {
  available_tokens?: number;
  parent_reserve_tokens?: number;
  recovery_reserve_tokens?: number;
  worker_pool_tokens?: number;
  max_concurrent?: number;
  [key: string]: unknown;
}

export interface DelegationEventPayload {
  delegation_run_id?: string;
  topology?: string;
  task_count?: number;
  budget?: DelegationBudgetPayload;
  task_id?: string;
  generation?: number;
  agent_type?: string;
  child_session_id?: string;
  phase?: string;
  previous_phase?: string;
  status?: string;
  integration_status?: string;
  verification?: Record<string, unknown>;
  action?: string;
  dependencies?: string[];
  changed_files?: string[];
  report_count?: number;
  tokens_used?: number;
  duration_ms?: number;
  reason?: string;
  error?: string;
}

/**
 * Delegation events may arrive flattened over WebSocket or in the durable
 * fallback shape `{ type, payload, timestamp }`. Keep both forms optional so
 * older servers and replayed traces remain readable.
 */
interface WsDelegationEventBase extends EventEnvelope, DelegationEventPayload {
  payload?: DelegationEventPayload;
  timestamp?: string;
}

export interface WsDelegationPlannedEvent extends WsDelegationEventBase {
  type: "delegation_planned";
}

export interface WsDelegationTaskQueuedEvent extends WsDelegationEventBase {
  type: "delegation_task_queued";
}

export interface WsDelegationTaskStartedEvent extends WsDelegationEventBase {
  type: "delegation_task_started";
}

export interface WsDelegationTaskReportedEvent extends WsDelegationEventBase {
  type: "delegation_task_reported";
}

export interface WsDelegationTaskFailedEvent extends WsDelegationEventBase {
  type: "delegation_task_failed";
}

export interface WsDelegationTaskBlockedEvent extends WsDelegationEventBase {
  type: "delegation_task_blocked";
}

export interface WsDelegationTaskRetryingEvent extends WsDelegationEventBase {
  type: "delegation_task_retrying";
}

export interface WsDelegationPhaseChangedEvent extends WsDelegationEventBase {
  type: "delegation_phase_changed";
}

export interface WsDelegationIntegrationStartedEvent extends WsDelegationEventBase {
  type: "delegation_integration_started";
}

export interface WsDelegationIntegrationCompletedEvent extends WsDelegationEventBase {
  type: "delegation_integration_completed";
}

export interface WsDelegationVerificationStartedEvent extends WsDelegationEventBase {
  type: "delegation_verification_started";
}

export interface WsDelegationVerificationCompletedEvent extends WsDelegationEventBase {
  type: "delegation_verification_completed";
}

export interface WsDelegationSynthesisStartedEvent extends WsDelegationEventBase {
  type: "delegation_synthesis_started";
}

export interface WsDelegationCompletedEvent extends WsDelegationEventBase {
  type: "delegation_completed";
}

export interface WsDelegationBudgetExhaustedEvent extends WsDelegationEventBase {
  type: "delegation_budget_exhausted";
}

export type WsDelegationEvent =
  | WsDelegationPlannedEvent
  | WsDelegationTaskQueuedEvent
  | WsDelegationTaskStartedEvent
  | WsDelegationTaskReportedEvent
  | WsDelegationTaskFailedEvent
  | WsDelegationTaskBlockedEvent
  | WsDelegationTaskRetryingEvent
  | WsDelegationPhaseChangedEvent
  | WsDelegationIntegrationStartedEvent
  | WsDelegationIntegrationCompletedEvent
  | WsDelegationVerificationStartedEvent
  | WsDelegationVerificationCompletedEvent
  | WsDelegationSynthesisStartedEvent
  | WsDelegationCompletedEvent
  | WsDelegationBudgetExhaustedEvent;

// ── Approval ────────────────────────────────────────────────────────────

export interface WsApprovalRequiredEvent extends EventEnvelope {
  type: "approval_required";
  request_id: string;
  tool_name: string;
  params?: Record<string, unknown>;
  thought?: string;
  decision_reason?: string;
  tool_use_id?: string;
  permission_mode?: string;
  risk_level?: string;
  timestamp?: string;
  step?: number;
}

export interface WsApprovalTimeoutEvent extends EventEnvelope {
  type: "approval_timeout";
  request_id: string;
  timestamp?: string;
}

export interface WsApprovalResolvedEvent extends EventEnvelope {
  type: "approval_resolved";
  request_id: string;
  tool_name: string;
  decision: "allow_once" | "always_allow" | "deny" | string;
  note?: string;
  updated_input?: boolean;
  wait_ms?: number;
  timestamp?: string;
}

// ── Plan ────────────────────────────────────────────────────────────────

export interface WsPlanReadyEvent extends EventEnvelope {
  type: "plan_ready";
  plan_text?: string;
  contract?: Record<string, unknown> | null;
  revision?: number;
  max_revisions?: number;
  result?: { summary?: string; steps_taken?: number; total_tokens?: number };
  timestamp?: string;
  step?: number;
}

// ── Worktree ────────────────────────────────────────────────────────────

export interface WsWorktreeResolvedEvent extends EventEnvelope {
  type: "worktree_resolved";
  child_session_id: string;
  action: string;
  status: string;
  message?: string;
  timestamp?: string;
  step?: number;
}

export interface WsReviewUpdatedEvent extends EventEnvelope {
  type: "review_updated";
  job_id: string;
  status: string;
  task_states: Record<string, string>;
  finding_count: number;
  workspace_revision: string;
  timestamp?: string;
}

// ── Memory activity ─────────────────────────────────────────────────────

export interface WsMemoryRecallEvent extends EventEnvelope {
  type: "memory_recall";
  injected_count: number;
  candidate_count: number;
  omitted_count: number;
  top_names: string[];
  timestamp?: string;
}

export interface WsMemoryWrittenEvent extends EventEnvelope {
  type: "memory_written";
  name: string;
  description: string;
  source: string;
  confidence: number;
  timestamp?: string;
}

// ── Assistant text streaming ─────────────────────────────────────────────

export interface WsAssistantTextStartEvent extends EventEnvelope {
  type: "assistant_text_start";
  block_id: string;
  timestamp?: string;
}

export interface WsAssistantTextDeltaEvent extends EventEnvelope {
  type: "assistant_text_delta";
  text: string;
  block_id: string;
  timestamp?: string;
}

export interface WsAssistantTextEndEvent extends EventEnvelope {
  type: "assistant_text_end";
  block_id: string;
  timestamp?: string;
}

export interface WsAssistantTextAbortedEvent extends EventEnvelope {
  type: "assistant_text_aborted";
  block_id: string;
  reason: string;
  timestamp?: string;
}

// ── Run lifecycle ────────────────────────────────────────────────────────

export interface WsRunStartedEvent extends EventEnvelope {
  type: "run_started";
  run_id: string;
  turn_id: string;
  turn_index: number;
  timestamp?: string;
}

export interface RunVerificationCheck {
  name: string;
  status: "passed" | "failed" | "skipped" | "unavailable" | string;
  command?: string;
  detail?: string;
  duration_ms?: number;
}

export interface RunVerification {
  status: "not_applicable" | "verified" | "unverified" | "unavailable" | "failed" | string;
  reason: string;
  checks?: RunVerificationCheck[];
}

export interface RunWorkspaceDelta {
  has_changes?: boolean;
  changed_files?: string[];
  patch_available?: boolean;
  source?: "git" | "tool_journal" | string;
  is_run_scoped?: boolean;
}

export interface WsRunTerminalEvent extends EventEnvelope {
  type: "run_terminal";
  run_id: string;
  turn_id: string;
  turn_index: number;
  status: "completed" | "failed" | "cancelled";
  summary: string;
  steps_taken: number;
  total_tokens: number;
  error?: string;
  termination_reason?: string;
  verification_status?: string;
  verification_reason?: string;
  verification?: RunVerification;
  workspace_delta?: RunWorkspaceDelta;
  timestamp?: string;
}

// ── Discriminated union ─────────────────────────────────────────────────

export type WsMessage =
  | WsStatusEvent
  | WsThoughtEvent
  | WsThoughtDeltaEvent
  | WsReflectionEvent
  | WsToolCallEvent
  | WsObservationEvent
  | WsSubagentStartEvent
  | WsSubagentStopEvent
  | WsDelegationEvent
  | WsApprovalRequiredEvent
  | WsApprovalTimeoutEvent
  | WsApprovalResolvedEvent
  | WsPlanReadyEvent
  | WsWorktreeResolvedEvent
  | WsReviewUpdatedEvent
  | WsMemoryRecallEvent
  | WsMemoryWrittenEvent
  | WsAssistantTextStartEvent
  | WsAssistantTextDeltaEvent
  | WsAssistantTextEndEvent
  | WsAssistantTextAbortedEvent
  | WsRunStartedEvent
  | WsRunTerminalEvent;

// ── Typed handler utility ───────────────────────────────────────────────

/** Narrow a WsMessage to a specific subtype. */
export type WsMessageOfType<T extends WsMessage["type"]> = Extract<WsMessage, { type: T }>;

// ── Transport-level envelope ────────────────────────────────────────────

/**
 * WsMessage with optional trace sequence number injected by the backend
 * transport layer (EventBus WebSocket broadcast + /timeline REST).
 *
 * ``seq`` is NOT a domain property of individual event types — it is
 * a per-session monotonic counter that the backend stamps on every
 * event at persist/broadcast time.  Use this type in store handlers
 * and transport adapters; render components should use ``WsMessage``.
 */
export type WsEnvelope = WsMessage & { seq?: number };

File: server/events.py
"""
server/events.py — typed WS event dataclasses.

Single source of truth for all WebSocket message shapes.
Replace the ad-hoc dict construction in _translate_event()
with these structured types.

The frontend mirrors these types in web/src/types/events.ts.
"""

from __future__ import annotations

from dataclasses import dataclass, field, asdict
from typing import Any, Literal


def _to_dict(obj) -> dict:
    """Serialize a dataclass to dict.

    Only skip None values and the 'type' discriminator is always included.
    Empty strings and empty containers are preserved — the frontend may
    depend on their presence (e.g. ev.error || 'default').
    """
    result = {}
    for k, v in asdict(obj).items():
        if v is None:
            continue
        result[k] = v
    return result


# ── Status events ─────────────────────────────────────────────────────


@dataclass
class WsStatus:
    type: Literal["status"] = "status"
    status: str = ""            # running | completed | failed | finish | gave_up | cancelled | compacted
    message: str = ""
    error: str = ""
    result: dict | None = None  # {summary, steps_taken, total_tokens}
    content: str = ""           # assistant 最终回答正文（由 task_complete.summary 填充）
    timestamp: str = ""
    # ── EventEnvelope fields ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Thought ───────────────────────────────────────────────────────────


@dataclass
class WsThought:
    type: Literal["thought"] = "thought"
    content: str = ""
    step: int = 0
    child_session_id: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsThoughtDelta:
    """Streaming thought token — pushed in real-time during LLM generation."""
    type: Literal["thought_delta"] = "thought_delta"
    text: str = ""
    step: int = 0
    child_session_id: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Tool call ─────────────────────────────────────────────────────────


@dataclass
class WsToolCall:
    type: Literal["tool_call"] = "tool_call"
    name: str = ""
    params: dict = field(default_factory=dict)
    step: int = 0
    id: str = ""
    child_session_id: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""       # = tool_call_id
    tool_call_id: str = ""   # model-assigned tool use id

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Observation ───────────────────────────────────────────────────────


@dataclass
class WsObservation:
    type: Literal["observation"] = "observation"
    tool_name: str = ""
    output: str = ""
    error: str = ""
    status: str = ""
    step: int = 0
    id: str = ""
    diff: str = ""
    evidence: dict | None = None
    """Evidence metadata: {evidence_id, kind, status, cached, source_fingerprint}"""
    child_session_id: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""       # NOT created — observation updates tool_use block
    tool_call_id: str = ""   # matches tool_call's id

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsEvidenceRecord:
    """One persisted evidence row projected to trace/WS after DB commit."""

    type: Literal["evidence_record"] = "evidence_record"
    evidence: dict = field(default_factory=dict)
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Reflection ────────────────────────────────────────────────────────


@dataclass
class WsReflection:
    type: Literal["reflection"] = "reflection"
    content: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Subagent events ───────────────────────────────────────────────────


@dataclass
class WsSubagentStart:
    type: Literal["subagent_start"] = "subagent_start"
    child_session_id: str = ""
    agent_name: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsSubagentStop:
    type: Literal["subagent_stop"] = "subagent_stop"
    child_session_id: str = ""
    status: str = ""
    timestamp: str = ""
    # ── EventEnvelope ──
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Approval events ───────────────────────────────────────────────────


@dataclass
class WsApprovalRequired:
    type: Literal["approval_required"] = "approval_required"
    request_id: str = ""
    tool_name: str = ""
    params: dict = field(default_factory=dict)
    thought: str = ""
    decision_reason: str = ""
    tool_use_id: str = ""
    permission_mode: str = ""
    risk_level: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsApprovalTimeout:
    type: Literal["approval_timeout"] = "approval_timeout"
    request_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsApprovalResolved:
    """Persisted response paired with one approval_required event."""

    type: Literal["approval_resolved"] = "approval_resolved"
    request_id: str = ""
    tool_name: str = ""
    decision: str = ""
    note: str = ""
    updated_input: bool = False
    wait_ms: float = 0.0
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Assistant text streaming ───────────────────────────────────────────


@dataclass
class WsAssistantTextStart:
    """Start of an assistant text block — creates a new text block with block_id."""
    type: Literal["assistant_text_start"] = "assistant_text_start"
    block_id: str = ""
    timestamp: str = ""
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsAssistantTextDelta:
    """Streaming token of assistant text — appends to the block identified by block_id."""
    type: Literal["assistant_text_delta"] = "assistant_text_delta"
    text: str = ""
    block_id: str = ""
    timestamp: str = ""
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsAssistantTextEnd:
    """End of an assistant text block — marks it as completed."""
    type: Literal["assistant_text_end"] = "assistant_text_end"
    block_id: str = ""
    timestamp: str = ""
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsAssistantTextAborted:
    """Text block was aborted — stream error, cancel, max_tokens, etc."""
    type: Literal["assistant_text_aborted"] = "assistant_text_aborted"
    block_id: str = ""
    reason: str = ""
    timestamp: str = ""
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Multi-agent delegation events ────────────────────────────────────


DelegationEventType = Literal[
    "delegation_planned",
    "delegation_task_queued",
    "delegation_task_started",
    "delegation_task_reported",
    "delegation_task_failed",
    "delegation_task_blocked",
    "delegation_task_retrying",
    "delegation_synthesis_started",
    "delegation_phase_changed",
    "delegation_integration_started",
    "delegation_integration_completed",
    "delegation_verification_started",
    "delegation_verification_completed",
    "delegation_completed",
    "delegation_budget_exhausted",
]


@dataclass
class WsDelegationEvent:
    """Typed, flattened delegation lifecycle event used by live WS and replay."""

    type: DelegationEventType = "delegation_phase_changed"
    delegation_run_id: str = ""
    task_id: str = ""
    generation: int = 0
    topology: str = ""
    task_count: int = 0
    phase: str = ""
    previous_phase: str = ""
    status: str = ""
    agent_type: str = ""
    child_session_id: str = ""
    report_count: int = 0
    tokens_used: int = 0
    duration_ms: int = 0
    reason: str = ""
    error: str = ""
    action: str = ""
    integration_status: str = ""
    verification: dict = field(default_factory=dict)
    budget: dict = field(default_factory=dict)
    changed_files: list[str] = field(default_factory=list)
    dependencies: list[str] = field(default_factory=list)
    timestamp: str = ""
    session_id: str = ""
    run_id: str = ""
    turn_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Run lifecycle events ───────────────────────────────────────────────


@dataclass
class WsRunStarted:
    """Run has transitioned QUEUED → RUNNING."""
    type: Literal["run_started"] = "run_started"
    run_id: str = ""
    turn_id: str = ""
    turn_index: int = 0
    timestamp: str = ""
    session_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsRunTerminal:
    """Run has reached a terminal state (completed/failed/cancelled).

    Sent only AFTER the final assistant message and Run status have been
    committed to the database.  This is the signal that the frontend can
    safely refresh from the DB.
    """
    type: Literal["run_terminal"] = "run_terminal"
    run_id: str = ""
    turn_id: str = ""
    turn_index: int = 0
    status: str = ""          # completed | failed | cancelled
    summary: str = ""
    steps_taken: int = 0
    total_tokens: int = 0
    error: str = ""
    termination_reason: str = ""
    verification_status: str = ""
    verification_reason: str = ""
    verification: dict = field(default_factory=dict)
    workspace_delta: dict = field(default_factory=dict)
    evidence_summary: dict = field(default_factory=dict)
    timestamp: str = ""
    session_id: str = ""
    event_id: str = ""
    sequence: int = 0
    block_id: str = ""
    tool_call_id: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Plan ready ────────────────────────────────────────────────────────


@dataclass
class WsPlanReady:
    type: Literal["plan_ready"] = "plan_ready"
    plan_text: str = ""
    contract: dict | None = None
    revision: int = 0
    max_revisions: int = 5
    result: dict | None = None
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Worktree resolved ─────────────────────────────────────────────────


@dataclass
class WsWorktreeResolved:
    type: Literal["worktree_resolved"] = "worktree_resolved"
    child_session_id: str = ""
    action: str = ""
    status: str = ""
    message: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsReviewUpdated:
    """A durable multi-agent review job or task changed state."""

    type: Literal["review_updated"] = "review_updated"
    job_id: str = ""
    status: str = ""
    task_states: dict[str, str] = field(default_factory=dict)
    finding_count: int = 0
    workspace_revision: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Memory activity ───────────────────────────────────────────────────


@dataclass
class WsMemoryRecall:
    type: Literal["memory_recall"] = "memory_recall"
    injected_count: int = 0
    candidate_count: int = 0
    omitted_count: int = 0
    top_names: list[str] = field(default_factory=list)
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsMemoryWritten:
    type: Literal["memory_written"] = "memory_written"
    name: str = ""
    description: str = ""
    source: str = ""
    confidence: float = 0.0
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Discriminated union ───────────────────────────────────────────────

WsEvent = (
    WsStatus | WsThought | WsThoughtDelta | WsToolCall | WsObservation | WsReflection
    | WsSubagentStart | WsSubagentStop
    | WsApprovalRequired | WsApprovalTimeout | WsApprovalResolved | WsPlanReady
    | WsWorktreeResolved | WsMemoryRecall | WsMemoryWritten
    | WsAssistantTextStart | WsAssistantTextDelta | WsAssistantTextEnd | WsAssistantTextAborted
    | WsRunStarted | WsRunTerminal
)
